streams/src/main/java/org/apache/kafka/streams/processor/api/ProcessorContext.java

```diff
@@ -55,6 +52,16 @@
      */
     TaskId taskId();
 
+    /**
+     * The metadata of the record, if it is defined. Note that as long as the processor is
```

Comment:
(I'll offer an opinion here because I stressed the importance of it on the mailing list, though I do think this is already pretty good.  Feel free to take it or leave it, it's just a rewording that may or may not add clarity)

> Note that as long as the processor is receiving a record downstream of a Source

Isn't every record received downstream of a source one way or another?  IMO the up front emphasis should be on _how_ the record came from the upstream processor.  My attempt:

> The record metadata is defined if the record currently being processed was passed through parent processor(s) directly from a Source.  It is undefined if the record was forwarded from within a Punctuator.

Reply:
Thanks, @pgwhalen !

I agree, the docs I wrote were a little too terse. How about this:

```java
    /**
     * The metadata of the source record, if is one. Processors may be invoked to
     * process a source record from an input topic, to run a scheduled punctuation
     * (see {@link ProcessorContext#schedule(Duration, PunctuationType, Punctuator)} ),
     * or because a parent processor called {@link ProcessorContext#forward(Record)}.
     * <p>
     * In the case of a punctuation, there is no source record, so this metadata would be
     * undefined. Note that when a punctuator invokes {@link ProcessorContext#forward(Record)},
     * downstream processors will receive the forwarded record as a regular
     * {@link Processor#process(Record)} invocation. In other words, it wouldn't be apparent to
     * downstream processors whether or not the record being processed came from an input topic
     * or punctuation and therefore whether or not this metadata is defined. This is why
     * the return type of this method is {@link Optional}.
     * <p>
     * If there is any possibility of punctuators upstream, any access
     * to this field should consider the case of
     * "{@code recordMetadata().isPresent() == false}".
     * Of course, it would be safest to always guard this condition.
     */
```

Reply:
I've just updated the PR with this change.